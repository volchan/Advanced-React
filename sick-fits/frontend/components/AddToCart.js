import React from "react";
import { Mutation } from "react-apollo";
import gql from "graphql-tag";

const ADD_TO_CART_MUTATION = gql`
  mutation ADD_TO_CART_MUTATION($id: ID!) {
    addToCart(id: $id) {
      id
      quantity
    }
  }
`;

class AddToCart extends React.Component {
  render() {
    return (
      <Mutation
        mutation={ADD_TO_CART_MUTATION}
        variables={{ id: this.props.id }}
      >
        {(addToCart, { loading }) => {
          return (
            <button onClick={addToCart}>
              Add
              {loading ? "ing" : ""} To Cart ðŸ›’
            </button>
          );
        }}
      </Mutation>
    );
  }
}

export default AddToCart;
