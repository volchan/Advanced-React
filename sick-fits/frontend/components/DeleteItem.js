import React, { Component } from "react";
import { Mutation } from "react-apollo";
import gql from "graphql-tag";

import { ALL_ITEMS_QUERY } from "./Items";

const DELETE_ITEM_MUTATION = gql`
  mutation DELETE_ITEM_MUTATION($id: ID!) {
    deleteItem(id: $id) {
      id
    }
  }
`;

class DeleteItem extends Component {
  update = (cache, payload) => {
    // manually update the cache, so it matches the server
    // read the cache for the items we want
    const data = cache.readQuery({ query: ALL_ITEMS_QUERY });
    // filter the delete item out of the page
    const filteredItems = data.items.filter(
      item => item.id !== payload.data.deleteItem.id
    );
    // put the items back
    cache.writeQuery({
      query: ALL_ITEMS_QUERY,
      data: { ...data, items: filteredItems }
    });
  };

  render() {
    return (
      <Mutation
        mutation={DELETE_ITEM_MUTATION}
        variables={{ id: this.props.id }}
        update={this.update}
      >
        {(deleteItem, { error }) => {
          return (
            <button
              onClick={() => {
                if (confirm("Are you sure you want to delete this item?")) {
                  deleteItem();
                }
              }}
            >
              {this.props.children}
            </button>
          );
        }}
      </Mutation>
    );
  }
}

export default DeleteItem;
