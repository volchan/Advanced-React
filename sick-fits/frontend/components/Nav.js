import Link from "next/link";

import NavStyles from "./styles/NavStyles";
import User from "./User";

String.prototype.capitalize = function () {
  return this.charAt(0).toUpperCase() + this.slice(1);
}

const Nav = () => (
  <NavStyles>
    <User>
      {({data: { me }}) => {
        if (me) return <p>{me.firstName.capitalize()} {me.lastName.toUpperCase()}</p>
        return null;
      }}
    </User>
    <Link href="/items"><a>Shop</a></Link>
    <Link href="/sell"><a>Sell</a></Link>
    <Link href="/signup"><a>Signup</a></Link>
    <Link href="/orders"><a>Orders</a></Link>
    <Link href="/me"><a>Account</a></Link>
  </NavStyles>
);

export default Nav;
