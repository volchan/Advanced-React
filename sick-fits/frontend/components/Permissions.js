import { Query } from "react-apollo";
import gql from "graphql-tag";

import Error from "./ErrorMessage";
import Table from "./styles/Table";
import SickButton from "./styles/SickButton";

const possiblePermissions = [
  "ADMIN",
  "USER",
  "ITEMCREATE",
  "ITEMUPDATE",
  "ITEMDELETE",
  "PERMISSIONUPDATE"
];

const ALL_USERS_QUERY = gql`
  query ALL_USERS_QUERY {
    users {
      id
      firstName
      lastName
      email
      permissions
    }
  }
`;

const Permissions = props => {
  return (
    <Query query={ALL_USERS_QUERY}>
      {({ data, loading, error }) => {
        return (
          <div>
            <Error error={error} />
            <div>
              <h2>Manage Permissions</h2>
              <Table>
                <thead>
                  <tr>
                    <th>Name</th>
                    <th>Email</th>
                    {possiblePermissions.map(permission => {
                      return <th key={permission}>{permission}</th>;
                    })}
                    <th />
                  </tr>
                </thead>
                <tbody>
                  {data.users.map(user => {
                    return <User user={user} key={user.id} />;
                  })}
                </tbody>
              </Table>
            </div>
          </div>
        );
      }}
    </Query>
  );
};

class User extends React.Component {
  render() {
    const user = this.props.user;
    return (
      <tr>
        <td>
          {user.firstName.capitalize()} {user.lastName.toUpperCase()}
        </td>
        <td>{user.email}</td>
        {possiblePermissions.map(permission => {
          return (
            <td key={`${user.id}-permission-${permission}`}>
              <label htmlFor={`${user.id}-permission-${permission}`}>
                <input type="checkbox" name="" id="" />
              </label>
            </td>
          );
        })}
        <td>
          <SickButton>Update</SickButton>
        </td>
      </tr>
    );
  }
}

export default Permissions;
