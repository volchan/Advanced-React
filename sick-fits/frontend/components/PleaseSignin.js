import { Query } from "react-apollo";

import Signin from "./Signin";
import { CURRENT_USER_QUERY } from "./User";

const PleaseSignin = props => {
  return (
    <Query query={CURRENT_USER_QUERY}>
      {({ data, loading }) => {
        if (loading) return <p>Loading...</p>;
        if (!data.me) {
          return (
            <div>
              <p>Please Sign In Before Continuing</p>
              <Signin />
            </div>
          );
        }
        return props.children;
      }}
    </Query>
  );
};

export default PleaseSignin;
