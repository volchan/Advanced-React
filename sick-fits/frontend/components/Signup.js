import React, { Component } from "react";
import { Mutation } from "react-apollo";
import gql from "graphql-tag";

import Form from "./styles/Form";
import Error from "./ErrorMessage";

const SIGNUP_MUTATION = gql`
  mutation SIGNUP_MUTATION(
    $email: String!
    $firstName: String!
    $lastName: String!
    $password: String!
  ) {
    signup(
      email: $email
      firstName: $firstName
      lastName: $lastName
      password: $password
    ) {
      id
      email
      firstName
      lastName
    }
  }
`;

class Signup extends Component {
  state = {
    email: "",
    firstName: "",
    lastName: "",
    password: ""
  };

  saveToState = e => {
    this.setState({ [e.target.name]: e.target.value });
  };

  render() {
    return (
      <Mutation mutation={SIGNUP_MUTATION} variables={this.state}>
        {(signup, { error, loading }) => {
          return (
            <Form
              method="post"
              onSubmit={async e => {
                e.preventDefault();
                await signup();
                this.setState({
                  email: "",
                  firstName: "",
                  lastName: "",
                  password: ""
                });
              }}
            >
              <fieldset disabled={loading} aria-busy={loading}>
                <h2>Sign Up for an Account</h2>
                <Error error={error} />
                <label htmlFor="email">
                  Email
                  <input
                    type="email"
                    name="email"
                    placeholder="Email"
                    value={this.state.email}
                    onChange={this.saveToState}
                  />
                </label>
                <label htmlFor="firstName">
                  First Name
                  <input
                    type="text"
                    name="firstName"
                    placeholder="First Name"
                    value={this.state.firstName}
                    onChange={this.saveToState}
                  />
                </label>
                <label htmlFor="lastName">
                  Last Name
                  <input
                    type="text"
                    name="lastName"
                    placeholder="Last Name"
                    value={this.state.lastName}
                    onChange={this.saveToState}
                  />
                </label>
                <label htmlFor="password">
                  Password
                  <input
                    type="password"
                    name="password"
                    placeholder="Password"
                    value={this.state.password}
                    onChange={this.saveToState}
                  />
                </label>
                <button type="submit">Sign Up!</button>
              </fieldset>
            </Form>
          );
        }}
      </Mutation>
    );
  }
}

export default Signup;
