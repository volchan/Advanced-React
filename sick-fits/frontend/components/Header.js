import Nav from "./Nav";

const Header = () => {
  return (
    <div>
        <div className="bar">
            <a href="">Sick Fits</a>
            <Nav />
        </div>
        <div className="sub-bar">
            <p>Search</p>
            <p>Cart</p>
        </div>
    </div>
  )
}

export default Header;
