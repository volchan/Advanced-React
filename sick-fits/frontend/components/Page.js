import React, { Component } from "react";
import styled from "styled-components";

import Header from "./Header";
import Meta from "./Meta";

const MyButton = styled.button`
  background: #00a1ff;
  padding: 5px;
  font-size: 1.5rem;
  color: white;
  border: 2px solid rgba(255, 255, 255, .25);
  /* font-size: ${props => (props.huge ? "100px" : "50px")}; */
  border-radius: 4px;
`;

export default class Page extends Component {
  render() {
    return <div>
      <Meta />
      <Header />
      <MyButton >
        Click Me
      </MyButton>
      {this.props.children}
    </div>;
  }
}
