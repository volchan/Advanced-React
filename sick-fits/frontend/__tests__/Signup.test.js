import { mount } from "enzyme";
import toJSON from "enzyme-to-json";
import wait from "waait";
import { MockedProvider } from "react-apollo/test-utils";
import { ApolloConsumer } from "react-apollo";

import Signup, { SIGNUP_MUTATION } from "../components/Signup";
import { CURRENT_USER_QUERY } from "../components/User";
import { fakeUser } from "../lib/testUtils";

const me = fakeUser();
const mocks = [
  {
    request: {
      query: SIGNUP_MUTATION,
      variables: {
        email: me.email,
        firstName: me.firstName,
        lastName: me.lastName,
        password: "password"
      }
    },
    result: {
      data: {
        signup: {
          ...me
        }
      }
    }
  },
  {
    request: {
      query: CURRENT_USER_QUERY
    },
    result: { data: { me } }
  }
];

const type = (wrapper, name, value) => {
  wrapper.find(`input[name="${name}"]`).simulate("change", {
    target: { name, value }
  });
};

describe("<Signup />", () => {
  it("renders ans matches snapshot", () => {
    const wrapper = mount(
      <MockedProvider>
        <Signup />
      </MockedProvider>
    );

    const form = wrapper.find('form[data-test="form"]');
    expect(toJSON(form)).toMatchSnapshot();
  });

  it("calls the mutation properly", async () => {
    let apolloClient;
    const wrapper = mount(
      <MockedProvider mocks={mocks}>
        <ApolloConsumer>
          {client => {
            apolloClient = client;
            return <Signup />;
          }}
        </ApolloConsumer>
      </MockedProvider>
    );

    await wait();
    type(wrapper, "firstName", me.firstName);
    type(wrapper, "lastName", me.lastName);
    type(wrapper, "email", me.email);
    type(wrapper, "password", "password");
    wrapper.update();
    wrapper.find('form[data-test="form"]').simulate("submit");

    await wait();
    const user = await apolloClient.query({ query: CURRENT_USER_QUERY });
    expect(user.data.me).toMatchObject(me);
  });
});
