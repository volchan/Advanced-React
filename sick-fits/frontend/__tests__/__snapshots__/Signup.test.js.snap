// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Signup /> renders ans matches snapshot 1`] = `
<form
  className="Form-sc-1xszr8q-0 krRxky"
  data-test="form"
  method="post"
  onSubmit={[Function]}
>
  <fieldset
    aria-busy={false}
    disabled={false}
  >
    <h2>
      Sign Up for an Account
    </h2>
    <DisplayError
      error={Object {}}
    />
    <label
      htmlFor="email"
    >
      Email
      <input
        name="email"
        onChange={[Function]}
        placeholder="Email"
        type="email"
        value=""
      />
    </label>
    <label
      htmlFor="firstName"
    >
      First Name
      <input
        name="firstName"
        onChange={[Function]}
        placeholder="First Name"
        type="text"
        value=""
      />
    </label>
    <label
      htmlFor="lastName"
    >
      Last Name
      <input
        name="lastName"
        onChange={[Function]}
        placeholder="Last Name"
        type="text"
        value=""
      />
    </label>
    <label
      htmlFor="password"
    >
      Password
      <input
        name="password"
        onChange={[Function]}
        placeholder="Password"
        type="password"
        value=""
      />
    </label>
    <button
      type="submit"
    >
      Sign Up!
    </button>
  </fieldset>
</form>
`;
